src/ray/pubsub/publisher.cc

```diff
@@ -246,6 +243,22 @@ std::unique_ptr<EntityState> SubscriptionIndex::CreateEntityState() {
 void SubscriberState::ConnectToSubscriber(const rpc::PubsubLongPollingRequest &request,
                                           rpc::PubsubLongPollingReply *reply,
                                           rpc::SendReplyCallback send_reply_callback) {
+  auto max_processed_sequence_id = request.max_processed_sequence_id();
+  if (request.publisher_id().empty() ||
+      publisher_id_ != PublisherID::FromBinary(request.publisher_id())) {
+    // in case the publisher_id mismatches, we should ignore the
+    // max_processed_sequence_id.
+    max_processed_sequence_id = 0;
+  }
+
+  // clean up messages that have already been processed.
+  while (!mailbox_.empty() &&
+         mailbox_.front()->sequence_id() <= max_processed_sequence_id) {
```

Comment:
I might be missing something, but isn't it going to cause issues if this happens?

1. New GCS starts. It sent messages to other nodes (so the global seq_no > 0)
2. publisher finds the subscriber doesn't match. Set max_processed_sequence_id for this subscriber == 0
2. `Publish` is called to this subscriber. Since seq_no > 0, it will never be sent (because max_processed_sequence_id == 0)?

Reply:
hmm

> Publish is called to this subscriber. Since seq_no > 0, it will never be sent (because max_processed_sequence_id == 0)?

on sender:  max_processed_sequence_id is only used for garbage collection, not preventing messaging being sent
on receiver: max_processed_sequence_id will also be reset to 0 so if any message sequence_id > 0, it will be received.